+nosnoc/+discrete_time_problem/Stewart.m

```diff
@@ -376,29 +410,49 @@ function generate_direct_transcription_constraints(obj)
-                    if opts.time_optimal_problem
-                        if opts.use_speed_of_time_variables
-                            obj.g.equidistant_control_grid(ii) = {[sum_h - opts.h;s_sot*sum_h - obj.w.T_final()/opts.N_stages]};
-                        else
-                            obj.g.equidistant_control_grid(ii) = {sum_h - obj.w.T_final()/opts.N_stages};
-                        end
+                %              to understand something about the time-freezing reformulation which is ugly.
+                % handle numerical time
+                if opts.use_fesd && opts.equidistant_control_grid
+                    relax = vdx.RelaxationStruct(opts.relax_terminal_numerical_time.to_vdx, 's_numerical_time', 'rho_numerical_time');
+                    if opts.time_optimal_problem && ~opts.time_freezing
+                        ecg_rhs = obj.w.T_final()/opts.N_stages;
                     else
-                        if opts.relax_terminal_numerical_time
-                            % Negative values sometimes help convergence.
-                            obj.w.s_numerical_time(ii) = {{'s_numerical', 1}, -2*opts.h, 2*opts.h, opts.h/2};
-                            g_eq_grid = [sum_h - t_stage - obj.w.s_numerical_time(ii);
-                                -(sum_h - t_stage) - obj.w.s_numerical_time(ii)];
-                            obj.g.equidistant_control_grid(ii) = {g_eq_grid, -inf, 0};
-                            obj.f = obj.f + opts.rho_terminal_numerical_time*obj.w.s_numerical_time(ii);
+                        ecg_rhs = obj.p.T()/opts.N_stages;
+                    end
+
+                    if opts.use_numerical_clock_state
+                        curr_t = obj.w.numerical_time(ii,opts.N_finite_elements(ii));
+                        obj.g.equidistant_numerical_grid(ii) = {curr_t - ii*ecg_rhs, relax};
+                    else
+                        if ~opts.time_freezing
+                            obj.g.equidistant_numerical_grid(ii) = {s_sot*sum_h - ecg_rhs, relax};
                         else
-                            obj.g.equidistant_control_grid(ii) = {t_stage-sum_h};
+                            obj.g.equidistant_numerical_grid(ii) = {sum_h - ecg_rhs, relax};
                         end
                     end
+                    if relax.is_relaxed
+                        obj.p.rho_numerical_time().val = opts.rho_terminal_numerical_time;
+                    end
+                end
+
+                % Handle possible physical time
+                if opts.time_freezing && opts.stagewise_clock_constraint
+                    x0 = obj.w.x(0,0,opts.n_s);
+                    t0 = x0(end);
+                    x_stage_end = obj.w.x(ii, opts.N_finite_elements(ii), opts.n_s+rbp);
+                    t_stage_end = x_stage_end(end);
+
+                    relax = vdx.RelaxationStruct(opts.relax_terminal_physical_time.to_vdx, 's_physical_time', 'rho_physical_time');
```

Comment:
as in heaviside, explain the vdx specific syntax here

Reply:
Explained in one comment in each Stewart and Heaviside.